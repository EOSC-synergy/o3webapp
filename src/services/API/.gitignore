# .gitignore in generated-client gets overwritten every run, thus putting this here instead
generated-client/.npmignore
generated-client/git_push.sh
generated-client/.openapi-generator-ignore
generated-client/.openapi-generator
openapitools.json
# gitignore the spec?
swagger.yml
