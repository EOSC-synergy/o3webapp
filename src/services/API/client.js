import axios from 'axios';
import { NO_MONTH_SELECTED } from '../../utils/constants';

/** The base URL */
const baseURL = "https://api.o3as.fedcloud.eu/api/v1";

/** The timeout value at which an error is thrown and fetching data stops in milliseconds */
const timeoutVal = 5 * 60 * 1000; // 5 min at least (fetching the models took 29s)

/**
 * Makes a GET request.
 *
 * @param {string} endpoint the endpoint of the URL
 * @returns the request promise from axios
 */
const getFromAPI = (endpoint) => {
    return axios.get(baseURL + endpoint, { timeout: timeoutVal });
}

/**
 * Makes a POST request.
 *
 * @param {string} endpoint the endpoint of the URL
 * @param {Array.<Object>} data the data for the params of the post request
 * @returns the request promise from axios
 */
const postAtAPI = (endpoint, data) => {
    // don't pack "data" in an object, the api accepts only an array of values (e.g. model list)
    return axios.post(baseURL + endpoint, data, { timeout: timeoutVal }); 
}

/**
 * Gets the plot types.
 *
 * @returns the request promise from axios
 */
export const getPlotTypes = () => {
  return getFromAPI("/plots");
};

/**
 * Gets the models.
 *
 * @param {string} plotType the plot type for which the models should be fetched
 * @param {string} select a selection of specific models
 * @returns the request promise from axios
 */
export const getModels = (plotType, select) => {
    const hasPlotType = typeof plotType !== "undefined";
    const hasSelect = typeof select !== "undefined";
    const hasOne = hasPlotType || hasSelect;
    const hasBoth = hasPlotType && hasSelect;
    
    return getFromAPI(
        `/models${hasOne ? "?" : ""}${hasPlotType ? `ptype=${plotType}` : ""}${hasBoth ? "&" : ""}${hasSelect ? `select=${select}` : ""}`);
}

/**
 * Gets the plot style via a post request.
 *
 * @param {string} plotType the plot type for which the plot style should be fetched
 * @returns the request promise from axios
 */
export const postModelsPlotStyle = (plotType) => {
    return postAtAPI(
        '/models/plotstyle',
        {
            ptype: plotType
        }
    );
}

/**
 * Performs a request to /plots/plotId to fetch the plot data from the api formatted with the given parameters.
 * 
 * @param {string} plotId a string describing the plot - has to be the official plot name (e.g. tco3_zm)
 * @param {int} latMin specifies the minimum latitude
 * @param {int} latMax specifies specifying the maximum latitude
 * @param {Array.<int>} months represents the selected months
 * @param {Array.<String>} modelList lists the desired models
 * @param {int} startYear from which point the data should start
 * @param {int} endYear until which point the data is required
 * @param {string} refModel the reference model to "normalize the data"
 * @param {int} refYear the reference year to "normalize the data"
 * @returns the request promise from axios
 */
export const getPlotData = ({plotId, latMin, latMax, months, modelList, startYear, endYear, refModel, refYear}) => {
    if (months.length === 0) {
        return Promise.reject(new Error(NO_MONTH_SELECTED));
    }
    if (plotId === 'tco3_zm') {
        return postAtAPI(
            `/plots/${plotId}?begin=${startYear}&end=${endYear}&month=${months.join(",")}&lat_min=${latMin}&lat_max=${latMax}&ref_meas=${refModel}&ref_year=${refYear}`,
            modelList
        );
    } else if (plotId === 'tco3_return') {
        return postAtAPI(
            `/plots/${plotId}?month=${months.join(",")}&lat_min=${latMin}&lat_max=${latMax}&ref_meas=${refModel}&ref_year=${refYear}`,
            modelList
        );
    }
}