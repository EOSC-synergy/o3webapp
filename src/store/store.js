import { applyMiddleware, combineReducers, configureStore, createStore } from '@reduxjs/toolkit';
import plotReducer from './plotSlice/plotSlice';
import modelsReducer, { setModelsOfModelGroup } from './modelsSlice/modelsSlice';
import referenceReducer from './referenceSlice/referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { HYDRATE, createWrapper } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';

const bindMiddleware = (middleware) => {
    if (process.env.NODE_ENV !== 'production') {
        const { composeWithDevTools } = require('redux-devtools-extension');
        return composeWithDevTools(applyMiddleware(...middleware));
    }
    return applyMiddleware(...middleware);
};

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state, action) => {
    if (action.type === HYDRATE) {
        const nextState = {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return nextState;
    } else {
        return combinedReducer(state, action);
    }
};

let store;

const initStore = () => {
    store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return store;
};

export { store };

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = configureStore({
        reducer: combinedReducer,
    });
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
