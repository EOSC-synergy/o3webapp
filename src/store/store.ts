import { Action, combineReducers, configureStore, ThunkAction } from '@reduxjs/toolkit';
import plotReducer, { GlobalPlotState } from './plotSlice';
import modelsReducer, { GlobalModelState, setModelsOfModelGroup } from './modelsSlice';
import referenceReducer, { GlobalReferenceState } from './referenceSlice';
import apiReducer from '../services/API/apiSlice/apiSlice';
import { DEFAULT_MODEL_GROUP } from '../utils/constants';
import { createWrapper, HYDRATE } from 'next-redux-wrapper';
import thunkMiddleware from 'redux-thunk';
import { useDispatch, useStore } from 'react-redux';

type GlobalState =
    | undefined
    | (GlobalPlotState & GlobalModelState & GlobalReferenceState & { api: any });

const combinedReducer = combineReducers({
    plot: plotReducer,
    models: modelsReducer,
    reference: referenceReducer,
    api: apiReducer,
});

const reducer = (state: GlobalState, action: { type: string; payload: any }) => {
    if (action.type === HYDRATE) {
        // if (state.count.count) nextState.count.count = state.count.count; // preserve count value on client side navigation
        return {
            ...state, // use previous state
            ...action.payload, // apply delta from hydration
        };
    } else {
        return combinedReducer(state, action);
    }
};

//let store: ReturnType<typeof configureStore<GlobalState, >>;

const initStore = () => {
    //store = createStore(reducer, bindMiddleware([thunkMiddleware]));
    return configureStore({
        reducer,
        devTools: process.env.NODE_ENV !== 'production',
        middleware: (getDefaultMiddleware) => getDefaultMiddleware().concat(thunkMiddleware),
    });
};

export type AppStore = ReturnType<typeof initStore>;
export type AppState = ReturnType<AppStore['getState']>;
export type AppThunk<ReturnType = void> = ThunkAction<ReturnType, AppState, unknown, Action>;

export type AppDispatch = AppStore['dispatch'];
export const useAppDispatch = () => useDispatch<AppDispatch>(); // Export a hook that can be reused to resolve types

export const useAppStore = () => useStore<AppState>();

export const wrapper = createWrapper(initStore);

export function createTestStore() {
    const store = initStore();
    store.dispatch(setModelsOfModelGroup(DEFAULT_MODEL_GROUP)); // fill with default group
    return store;
}
