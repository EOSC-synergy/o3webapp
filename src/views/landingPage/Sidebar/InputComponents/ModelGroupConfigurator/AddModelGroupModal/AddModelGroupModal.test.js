import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import AddModelGroupModal from './AddModelGroupModal';
import { Provider } from "react-redux";
import * as redux from 'react-redux';
import { createTestStore } from '../../../../../../store/store';
import { REQUEST_STATE } from "../../../../../../services/API/apiSlice";
import { TestScheduler } from 'jest';

const TEST_MODEL_NAME = "CCMI-1_ACCESS_ACCESS-CCM-refC2";

let store;
describe('test addModelGroupModal rendering', () => {
    beforeEach(() => {
        store = createTestStore();
    });

    it('renders without crashing', () => {
        const mockFunction = jest.fn();
        render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={() => {}} reportError={() => {}} />
        </Provider>);
    });

    
    it('renders correctly when open', () => {
        let { baseElement, container } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={() => {}} reportError={() => {}} />
        </Provider>
        );
        expect(baseElement).toMatchSnapshot();
        expect(container).toBeVisible();
    });

    it('renders correctly when closed', () => {
        let { container, baseElement } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={false} onClose={() => {}} reportError={() => {}} />
        </Provider>
        );
        expect(baseElement).toMatchSnapshot();
        expect(container).not.toBeVisible;
    });

    it('fires props.onClose when clicking on the closing icon button', () => {
        const onClose = jest.fn();
        const { getByTestId, container } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={onClose} reportError={() => {}} />
        </Provider>
        );
        const closeButton = getByTestId(/close-button/);
        fireEvent.click(closeButton);
        expect(onClose).toHaveBeenCalled();
        expect(container).not.toBeVisible;
    });

    it('raises a console.error function if a required prop is not provided', () => {
        console.error = jest.fn();
        render(<Provider store={store}>
            <AddModelGroupModal />
        </Provider>);
        expect(console.error).toHaveBeenCalled();
        render(<Provider store={store}>
            <AddModelGroupModal isOpen={false} reportError={() => {}} />
        </Provider>);
        expect(console.error).toHaveBeenCalled();
        render(<Provider store={store}>
            <AddModelGroupModal onClose={() => {}} />
        </Provider>
        );
        expect(console.error).toHaveBeenCalled();
    });

    it("renders correctly if model list provided", () => {

        const { baseElement } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={() => {}} reportError={() => {}}/>
        </Provider>);

        expect(baseElement).toMatchSnapshot();
    });
});

describe('test addModelGroupModal functionality', () => {

    beforeEach(() => {
        store = createTestStore();
    });


    it('disables move all checked buttons if nothing is checked at beginning', () => {
        const { getByTestId } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={() => {}} reportError={() => {}} />
        </Provider>
        );
        const moveAllRightButton = getByTestId(/move-allChecked-left/);
        expect(moveAllRightButton).toBeDisabled;
        const moveAllLeftButton = getByTestId(/move-allChecked-right/);
        expect(moveAllLeftButton).toBeDisabled;
    });

    it("displays spinner when models are being fetched", () => {
        const { getAllByTestId } = render(<Provider store={store}>
            <AddModelGroupModal isOpen={true} onClose={() => {}} reportError={() => {}} />
        </Provider>);
        const circularProgress = getAllByTestId(/AddModelGroupModal-spinner/);
        expect(circularProgress.length).toBe(2);
    });

    test.todo("call props.reportError if fetching models failed");
    test.todo("check if models are rendered on the left");
    test.todo("check if models can be moved from left to right");
    test.todo("check if models not being selected by search are being hidden");
    test.todo("check default model group name is rendered when props.modelGroupId is provided");
    test.todo("check default models on the right are rendered when props.modelGroupId is provided");
    test.todo("check if onClosing the modal resets all model groups on the right and checked");
});
