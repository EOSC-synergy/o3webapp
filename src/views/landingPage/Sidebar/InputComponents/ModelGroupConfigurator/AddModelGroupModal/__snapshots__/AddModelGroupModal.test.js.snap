// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test addModelGroupModal rendering renders correctly if model list provided 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    data-testid="AddModelGroupModal-modal-wrapper"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-50j5vj-MuiPaper-root-MuiCard-root"
      tabindex="-1"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
          >
            Add Model Group
          </span>
        </div>
        <div
          class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="addModelGroupModal-close-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-go2vvz-MuiGrid-root"
          style="margin-bottom: 2em;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-3d5apm-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-describedby="mui-4-helper-text"
                aria-invalid="false"
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="mui-4"
                placeholder="Your group"
                type="text"
                value=""
              />
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium css-1d1r5q-MuiFormHelperText-root"
              id="mui-4-helper-text"
            >
              The name will only appear in the legend of the exported plot.
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1yuhvjn"
          id="modal-modal-description"
        >
          <div
            class="css-8xjr4f"
          >
            <div
              class="css-1iz4iln"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary css-rzandp-MuiInputBase-root"
            >
              <input
                alt="Searchbar"
                aria-label="search"
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                data-testid="SearchbarInput"
                placeholder="Searchâ€¦"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1dnxiqs-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 css-11lxi55-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                All Available Models
              </p>
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                data-testid="AddModelGroupModal-spinner-left"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
              >
                <button
                  aria-label="move selected right"
                  class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root Mui-disabled  css-ak74bx-MuiButtonBase-root-MuiButton-root"
                  data-testid="AddModelGroupModal-button-move-allChecked-right"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  &gt;
                </button>
                <button
                  aria-label="move selected left"
                  class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root Mui-disabled  css-ak74bx-MuiButtonBase-root-MuiButton-root"
                  data-testid="AddModelGroupModal-button-move-allChecked-left"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  &lt;
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 css-11lxi55-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                Models in 
                your group
              </p>
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                data-testid="AddModelGroupModal-spinner-right"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-19l6csj-MuiCardActions-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add group
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;

exports[`test addModelGroupModal rendering renders correctly when closed 1`] = `
<body
  style=""
>
  <div />
</body>
`;

exports[`test addModelGroupModal rendering renders correctly when open 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    aria-describedby="modal-modal-description"
    aria-labelledby="modal-modal-title"
    class="MuiModal-root css-79ws1d-MuiModal-root"
    data-testid="AddModelGroupModal-modal-wrapper"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root css-i9fmh8-MuiBackdrop-root-MuiModal-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-test="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-50j5vj-MuiPaper-root-MuiCard-root"
      tabindex="-1"
    >
      <div
        class="MuiCardHeader-root css-185gdzj-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content css-1qbkelo-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-h5 MuiCardHeader-title css-1qvr50w-MuiTypography-root"
          >
            Add Model Group
          </span>
        </div>
        <div
          class="MuiCardHeader-action css-sgoict-MuiCardHeader-action"
        >
          <button
            aria-label="close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
            data-testid="addModelGroupModal-close-button"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-go2vvz-MuiGrid-root"
          style="margin-bottom: 2em;"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-3d5apm-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiInput-root MuiInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-1480iag-MuiInputBase-root-MuiInput-root"
            >
              <input
                aria-describedby="mui-2-helper-text"
                aria-invalid="false"
                class="MuiInput-input MuiInputBase-input css-1x51dt5-MuiInputBase-input-MuiInput-input"
                id="mui-2"
                placeholder="Your group"
                type="text"
                value=""
              />
            </div>
            <p
              class="MuiFormHelperText-root MuiFormHelperText-sizeMedium css-1d1r5q-MuiFormHelperText-root"
              id="mui-2-helper-text"
            >
              The name will only appear in the legend of the exported plot.
            </p>
          </div>
        </div>
        <div
          class="MuiBox-root css-1yuhvjn"
          id="modal-modal-description"
        >
          <div
            class="css-8xjr4f"
          >
            <div
              class="css-1iz4iln"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary css-rzandp-MuiInputBase-root"
            >
              <input
                alt="Searchbar"
                aria-label="search"
                class="MuiInputBase-input css-yz9k0d-MuiInputBase-input"
                data-testid="SearchbarInput"
                placeholder="Searchâ€¦"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-1dnxiqs-MuiGrid-root"
          >
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 css-11lxi55-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                All Available Models
              </p>
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                data-testid="AddModelGroupModal-spinner-left"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-2 css-63b4os-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-t0zib5-MuiGrid-root"
              >
                <button
                  aria-label="move selected right"
                  class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root Mui-disabled  css-ak74bx-MuiButtonBase-root-MuiButton-root"
                  data-testid="AddModelGroupModal-button-move-allChecked-right"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  &gt;
                </button>
                <button
                  aria-label="move selected left"
                  class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeSmall MuiButton-outlinedSizeSmall MuiButtonBase-root Mui-disabled  css-ak74bx-MuiButtonBase-root-MuiButton-root"
                  data-testid="AddModelGroupModal-button-move-allChecked-left"
                  disabled=""
                  tabindex="-1"
                  type="button"
                >
                  &lt;
                </button>
              </div>
            </div>
            <div
              class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-5 css-11lxi55-MuiGrid-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                Models in 
                your group
              </p>
              <span
                class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
                data-testid="AddModelGroupModal-spinner-right"
                role="progressbar"
                style="width: 40px; height: 40px;"
              >
                <svg
                  class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
                  viewBox="22 22 44 44"
                >
                  <circle
                    class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
                    cx="44"
                    cy="44"
                    fill="none"
                    r="20.2"
                    stroke-width="3.6"
                  />
                </svg>
              </span>
            </div>
          </div>
        </div>
        
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-19l6csj-MuiCardActions-root"
      >
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Add group
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <div
      data-test="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
